File: Problems/155_Min_Stack/Min_Stack.py
import sys

class Node:
    def __init__(self, data):
        self.next = None
        self.data = data


class MinStack:    
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min = sys.maxsize
        self.head = None


    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x < self.min:
            self.min = x
        
        self.node = Node(min)
        self.node.next = Node(x)

        if self.head == None:
            self.head = self.node
        else:
            temp = self.head
            self.head = self.node
            self.head.next.next = temp


    def pop(self):
        """
        :rtype: void
        """
        if self.head == None:
            temp = temp
        else:
            currHead = self.head.next
            if currHead.next != None:
                self.head = currHead.next
                self.min = self.head.data
            else:
                self.head = None
                self.min = sys.maxsize


    def top(self):
        """
        :rtype: int
        """
        if self.head != None:
            if self.head.next != None:
                return self.head.next.data
        else:
            return 0


    def getMin(self):
        """
        :rtype: int
        """
        return self.min
        
    
def calc():
    minStack = MinStack()

    minStack.push(-2)
    print("push(-2)")

    minStack.push(0)
    print("push(0)")

    minStack.push(-3)
    print("push(-3)")

    print("getMin() => %s" %(minStack.getMin()))

    minStack.pop()
    print("pop()")

    minStack.top()
    print("pop()")

    print("getMin() => %s" %(minStack.getMin()))


def main():
    calc()

if __name__ == "__main__":
    main()

File: Problems/0155_Min_Stack/Project_CS/Min_Stack.cs
using System;

public class MinStack {
    int min = int.MaxValue;
    Node head;

    private class Node {
        public int data;
        public Node next;

        public Node(int data) {
            next = null;
            this.data = data;
        }
    }

    public MinStack() {

    }

    public void Push(int x) {
        if (x < min) {
            min = x;
        }
        Node node = new Node(min);
        node.next = new Node(x);

        if (this.head == null || head.next == null) {
            head = node;
        } else {
            Node temp = head;
            head = node;
            head.next.next = temp;
        }
    }

    public void Pop() {
        if (this.head == null) {
            // Do nothing
        } else {
            Node currHead = head.next;
            if (currHead.next != null) {
                head = currHead.next;
                min = head.data;
            } else {
                head = null;
                this.min = int.MaxValue;
            }
        }
    }

    public int Top() {
        if (head != null)
            return head.next.data;
        else
            return 0;
    }

    public int GetMin() {
        return min;

    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(x);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */
public class Solution
{
    private void calc()
    {
        MinStack minStack = new MinStack();
        int result;

        minStack.Push(-2);
        Console.WriteLine("Push(-2)");

        minStack.Push(0);
        Console.WriteLine("Push(0)");

        minStack.Push(-3);
        Console.WriteLine("Push(-3)");

        result = minStack.GetMin();
        Console.WriteLine("GetMin() --> " + result.ToString());

        minStack.Pop();
        Console.WriteLine("Pop())");

        result = minStack.Top();
        Console.WriteLine("Top()    --> " + result.ToString());

        result = minStack.GetMin();
        Console.WriteLine("GetMin() --> " + result.ToString());
    }

    public void Main()
    {
        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        calc();

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
